### DISCUSSION TITLE
Pattern for Error Messages from the server to the client.

### DISCUSSION THREAD

**Discussion by markdav-is** [oqtane/oqtane.framework#3946] [Q&A]
We try and do as much server  processing as possible in our apps.  For a given client action, we might have a form and when that form is posted to the server, we might call a multi-step process with various APIs.  If everything goes well, we'll get a nice viewmodel back.  For the cases when something goes works, say an API lookup, we will fail on the server.

When the server fails we can modify the return code and also supply content.  It's possible in ASP.Net to return 401 (not found) and also a message with more details.  In Oqtane, using ServiceBase, the content is ignored when there's an error.

I'm about to work around this by creating our own ServiceBase that changes the behavior to also pass the content when there's an error.

Am I missing about services base and returning messages from the server?

**Comment by sbwalker** [comment]
@markdav-is ServiceBase is a wrapper around HttpClient to provide methods which support multi-tenancy and reduce the amount of code required to make API calls. ServiceBase does have a CheckResponse method which checks the status of the API call and logs the result to the Oqtane error log if it is an error.

**Reply by markdav-is** [reply]
I see that and it's handy.  We use HTTP errors to signal things like 404-not found, but the server can also supply a message to the client, like Account not found or Product not found from a single controller call.  ServiceBase eats the error code and content so the only way you can tell there was an error is to see if the return value is the default value and "something went wrong" is about all you know.

One option is to stick with service base and move more complex logic into the client service.  Then take the single controller call and break it up so that each call only does one thing. that will allow the client service to react differently based on not working.

My other option is to extend service base to preserve and return the http response.  This is the way I'm leaning.

**Reply by sbwalker** [reply]
@markdav-is I understand what you saying now. You are correct that ServiceBase follows a pattern where the client UI components are abstracted from the HttpClient transport details. Basically, the "service layer" is responsible for interacting with the backend APIs and the interface it exposes to UI components typically includes only strongly typed objects. 

Another pattern which is sometimes used in Blazor is to simply call HttpClient directly from within your UI components. In this scenario there is no "service layer" abstraction so you have access to all of the request/response details within your component. However the disadvantage to this approach is there is no clean separation of concerns. 

You will notice that in the transition to Blazor Web in .NET 8, the "service layer" actually fulfills another purpose as well. Basically you can have 2 implementations of your service interface. One implementation uses HttpClient and is used when running a true Blazor client like WebAssembly or .NET MAUI. The other implementation runs directly on the server and is used in Static rendering or Interactive Rendering using SignalR. The application determines at run-time which service implementation should be used based on the scenario. This approach would not be possible if you used HttpClient directly within your UI components.

**Reply by markdav-is** [reply]
the final answer ended up being something like this:

```
        protected async Task<(T,HttpResponseMessage)> GetJsonResponseAsync<T>(string uri)
        {
            var response = await GetHttpClient().GetAsync(uri, HttpCompletionOption.ResponseHeadersRead, CancellationToken.None);
            if (await CheckResponse(response, uri) && ValidateJsonContent(response.Content))
            {
                return (await response.Content.ReadFromJsonAsync<T>(), response);
                
            }
            return (default(T), response);
        }
```
and then calling it like this:
```
    public async Task<PostPaymentResponse> GetStripeSessionAsync(string id, int visitorId)
    {

        var result = (Data: default(PostPaymentResponse), Response: default(HttpResponseMessage));
        result = await GetJsonResponseAsync<PostPaymentResponse>($"{Apiurl}/getstripesession/{id}/{visitorId}");

        if (!result.Response.IsSuccessStatusCode) {
            result.Data = new PostPaymentResponse() { 
                ErrorMessage = result.Response.Content.ReadAsStringAsync().Result
            };
        }

        return result.Data;
    }
```
